File: CS/NonPersistentObjectsDemo.Module/BusinessObjects/CollectionStoredSeparately/Epoch.cs
﻿using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Serialization;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Xpo;

namespace NonPersistentObjectsDemo.Module.BusinessObjects {

    [DevExpress.ExpressApp.DC.XafDefaultProperty(nameof(Epoch.Name))]
    [DefaultClassOptions]
    public class Epoch : BaseObject, IObjectSpaceLink {
        public Epoch(Session session) : base(session) { }

        private string _Name;
        public string Name {
            get { return _Name; }
            set { SetPropertyValue<string>(nameof(Name), ref _Name, value); }
        }

        #region Technologies
        private BindingList<Technology> _Technologies;
        public BindingList<Technology> Technologies {
            get {
                if(_Technologies == null) {
                    _Technologies = new BindingList<Technology>();
                }
                return _Technologies;
            }
        }
        private string _TechnologyList;
        [Browsable(false)]
        [Size(SizeAttribute.Unlimited)]
        public string TechnologyList {
            get { return _TechnologyList; }
            set { SetPropertyValue<string>(nameof(TechnologyList), ref _TechnologyList, value); }
        }
        #endregion

        protected override void OnLoaded() {
            base.OnLoaded();
            LoadList(Technologies, TechnologyList);
        }
        protected override void OnSaving() {
            TechnologyList = SaveList(Technologies);
            base.OnSaving();
        }
        private IObjectSpace _objectSpace;
        protected IObjectSpace ObjectSpace { get { return _objectSpace; } }
        IObjectSpace IObjectSpaceLink.ObjectSpace {
            get { return _objectSpace; }
            set {
                if(_objectSpace != value) {
                    _objectSpace = value;
                }
            }
        }

        #region NP Serialization
        private void LoadList(IList<Technology> list, string data) {
            list.Clear();
            if(data != null) {
                foreach(var s in data.Split(',')) {
                    Guid key;
                    if(Guid.TryParse(s, out key)) {
                        var obj = ObjectSpace.GetObjectByKey<Technology>(key);
                        if(obj != null) {
                            list.Add(obj);
                        }
                    }
                }
            }
        }
        private string SaveList(IList<Technology> list) {
            if(list == null || list.Count == 0) {
                return null;
            }
            return string.Join(",", list.Select(o => o.Oid.ToString("D")));
        }
        #endregion
    }
}

File: VB/NonPersistentObjectsDemo.Module/BusinessObjects/CollectionStoredSeparately/Epoch.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Xml.Serialization
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Xpo

Namespace NonPersistentObjectsDemo.Module.BusinessObjects

	<DevExpress.ExpressApp.DC.XafDefaultProperty(NameOf(Epoch.Name))>
	<DefaultClassOptions>
	Public Class Epoch
		Inherits BaseObject
		Implements IObjectSpaceLink

		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		Private _Name As String
		Public Property Name() As String
			Get
				Return _Name
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)(NameOf(Name), _Name, value)
			End Set
		End Property

		#Region "Technologies"
		Private _Technologies As BindingList(Of Technology)
		Public ReadOnly Property Technologies() As BindingList(Of Technology)
			Get
				If _Technologies Is Nothing Then
					_Technologies = New BindingList(Of Technology)()
				End If
				Return _Technologies
			End Get
		End Property
		Private _TechnologyList As String
		<Browsable(False)>
		<Size(SizeAttribute.Unlimited)>
		Public Property TechnologyList() As String
			Get
				Return _TechnologyList
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)(NameOf(TechnologyList), _TechnologyList, value)
			End Set
		End Property
		#End Region

		Protected Overrides Sub OnLoaded()
			MyBase.OnLoaded()
			LoadList(Technologies, TechnologyList)
		End Sub
		Protected Overrides Sub OnSaving()
			TechnologyList = SaveList(Technologies)
			MyBase.OnSaving()
		End Sub
		Private _objectSpace As IObjectSpace
		Protected ReadOnly Property ObjectSpace() As IObjectSpace
			Get
				Return _objectSpace
			End Get
		End Property
		Private Property IObjectSpaceLink_ObjectSpace() As IObjectSpace Implements IObjectSpaceLink.ObjectSpace
			Get
				Return _objectSpace
			End Get
			Set(ByVal value As IObjectSpace)
				If _objectSpace IsNot value Then
					_objectSpace = value
				End If
			End Set
		End Property

		#Region "NP Serialization"
		Private Sub LoadList(ByVal list As IList(Of Technology), ByVal data As String)
			list.Clear()
			If data IsNot Nothing Then
				For Each s In data.Split(","c)
					Dim key As Guid = Nothing
					If Guid.TryParse(s, key) Then
						Dim obj = ObjectSpace.GetObjectByKey(Of Technology)(key)
						If obj IsNot Nothing Then
							list.Add(obj)
						End If
					End If
				Next s
			End If
		End Sub
		Private Function SaveList(ByVal list As IList(Of Technology)) As String
			If list Is Nothing OrElse list.Count = 0 Then
				Return Nothing
			End If
			Return String.Join(",", list.Select(Function(o) o.Oid.ToString("D")))
		End Function
		#End Region
	End Class
End Namespace
